---
title: "Análisis de una secuencia (Task 1)"
author: "Alex-Alex-Helena"
date: '`r format(Sys.Date(),"%e %B, %Y")`'
output: 
  html_document:
    toc: false
    keep_md: no
  pdf_document: 
    toc: false
    latex_engine: xelatex
params:
  path: "/home/ajo/gitRepos/bioinformatics/Task1"
  name_fasta: "gi32141095_N_1.fa"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = "#>", warning = F)

# Installs packages that have not been installed on the system automatically. 
# https://stackoverflow.com/questions/9341635/check-for-installed-packages-before-running-install-packages
packages <- c("Biostrings")
install.packages(setdiff(packages, rownames(installed.packages())))  

# Import packages listed in the 'packages' variable above. 
# https://stackoverflow.com/questions/8175912/load-multiple-packages-at-once
lapply(packages, require, character.only = TRUE)
```

Este es el primer análisis dinámico de una secuencia.

El fichero de entrada es cualquier secuencia con formato fasta con un tamaño máximo de la secuencia de una linea. Por tanto, el fichero fasta ocupa siempre dos lineas.

La secuencia leida del fichero **`r params$name_fasta`** es:

```{r, echo = F}
# the pdf engine cannot render this output because of the characters in the DNA table. Not sure if this can be fixed. 
filepath <- paste(params$path, params$name_fasta, sep = "/")
DNA <- readDNAStringSet(filepath)
as.character(DNA)
#names(DNA)
#toString(DNA)
```
La longitud de la secuencia es de *`r width(DNA)`* bases. Estan repartidas como:

* **Adeninas** (A): `r vcountPattern("A", DNA)`
* **Citosinas** (C): `r vcountPattern("C", DNA)`
* **Guaninas** (G): `r vcountPattern("G", DNA)`
* **Timinas** (T): `r vcountPattern("T", DNA)`
